### ISSUE TITLE
Out of Memory in PW SDK 5.1.1

### ISSUE THREAD

**Issue by n-belokopytov** [Pushwoosh/pushwoosh-android-sdk#62]
Hello team. Here are some OOM exceptions with stacktraces that we have started to get recently. It appears that they also have something to do with scheduling of local notifications. We get these two crashes on all kinds of devices with all kinds of Android versions, without a clear driver.

> com.pushwoosh.a.f.a
java.lang.OutOfMemoryError: Failed to allocate a 131084 byte allocation with 7978296 free bytes and 7MB until OOM; failed due to fragmentation (required continguous free 135168 bytes where largest contiguous free 114688 bytes)

1 | java.util.HashMap.inflateTable | HashMap.java, line 287
-- | -- | --
2 | java.util.HashMap.<init> | HashMap.java, line 255
3 | android.app.SharedPreferencesImpl$EditorImpl.commitToMemory | SharedPreferencesImpl.java, line 402
4 | android.app.SharedPreferencesImpl$EditorImpl.apply | SharedPreferencesImpl.java, line 362
5 | com.pushwoosh.a.f.a | Unknown Source
6 | com.pushwoosh.notification.LocalNotificationReceiver.scheduleNotification | Unknown Source
7 | com.pushwoosh.notification.e.f | Unknown Source
8 | com.pushwoosh.f.a | Unknown Source
9 | com.pushwoosh.f.a | Unknown Source
10 | com.pushwoosh.i.onReceive | Unknown Source
11 | com.pushwoosh.internal.event.EventBus.b | Unknown Source
12 | com.pushwoosh.internal.event.EventBus.c | Unknown Source
13 | com.pushwoosh.internal.event.EventBus.a | Unknown Source
14 | com.pushwoosh.internal.event.c.run | Unknown Source
15 | android.os.Handler.handleCallback | Handler.java, line 751

> com.pushwoosh.internal.preference.PreferenceIntValue.set
java.lang.OutOfMemoryError: Failed to allocate a 24 byte allocation with 296 free bytes and 296B until OOM

1 | java.util.HashMap.constructorNewEntry | HashMap.java, line 490
-- | -- | --
2 | java.util.HashMap.constructorPut | HashMap.java, line 460
3 | java.util.HashMap.constructorPutAll | HashMap.java, line 209
4 | java.util.HashMap.<init> | HashMap.java, line 196
5 | android.app.SharedPreferencesImpl$EditorImpl.commitToMemory | SharedPreferencesImpl.java, line 402
6 | android.app.SharedPreferencesImpl$EditorImpl.apply | SharedPreferencesImpl.java, line 362
7 | com.pushwoosh.internal.preference.PreferenceIntValue.set | Unknown Source
8 | com.pushwoosh.a.f.c | Unknown Source
9 | com.pushwoosh.notification.LocalNotificationReceiver.scheduleNotification | Unknown Source
10 | com.pushwoosh.notification.e.f | Unknown Source
11 | com.pushwoosh.f.a | Unknown Source
12 | com.pushwoosh.f.a | Unknown Source
13 | com.pushwoosh.i.onReceive | Unknown Source
14 | com.pushwoosh.internal.event.EventBus.b | Unknown Source
15 | com.pushwoosh.internal.event.EventBus.c | Unknown Source
16 | com.pushwoosh.internal.event.EventBus.a | Unknown Source
17 | com.pushwoosh.internal.event.c.run | Unknown Source
18 | android.os.Handler.handleCallback

Please help us understand what is happening.

**Comment by wfhm** [comment]
Hi,

We are working on it right now, I will update this thread once there is a fix rolled out.

**Comment by n-belokopytov** [comment]
@wfhm hello, please give us an ETA.

**Comment by wfhm** [comment]
@n-belokopytov it is expected to be fixed with a mid-September release.

**Comment by n-belokopytov** [comment]
@wfhm so here are a couple of new OutOfMemory stacks for you, guys. We have PW 5.1.1 and Firebase 11.4.2

1) Happens on any device, any Android OS.
```
com.pushwoosh.internal.preference.PreferenceIntValue.set
java.lang.OutOfMemoryError: Failed to allocate a 262156 byte allocation with 2949944 free bytes and 2MB until OOM; failed due to fragmentation (required continguous free 266240 bytes where largest contiguous free 53248 bytes)


Exception Backtrace:
--
1 | java.util.HashMap.inflateTable | HashMap.java, line 287
2 | java.util.HashMap.<init> | HashMap.java, line 255
3 | android.app.SharedPreferencesImpl$EditorImpl.commitToMemory | SharedPreferencesImpl.java, line 402
4 | android.app.SharedPreferencesImpl$EditorImpl.apply | SharedPreferencesImpl.java, line 362
5 | com.pushwoosh.internal.preference.PreferenceIntValue.set | Unknown Source
6 | com.pushwoosh.a.f.c | Unknown Source
7 | com.pushwoosh.notification.LocalNotificationReceiver.scheduleNotification | Unknown Source
8 | com.pushwoosh.notification.e.f | Unknown Source
9 | com.pushwoosh.g.a | Unknown Source
10 | com.pushwoosh.g.a | Unknown Source
11 | com.pushwoosh.j.onReceive | Unknown Source
12 | com.pushwoosh.internal.event.EventBus.b | Unknown Source
13 | com.pushwoosh.internal.event.EventBus.c | Unknown Source
14 | com.pushwoosh.internal.event.EventBus.a | Unknown Source
15 | com.pushwoosh.internal.event.c.run | Unknown Source
16 | android.os.Handler.handleCallback

```

2) Samsung devices, any OS version.

```
com.pushwoosh.a.f.a
java.lang.OutOfMemoryError: Failed to allocate a 24 byte allocation with 4592744 free bytes and 4MB until OOM; failed due to fragmentation (required continguous free 4096 bytes for a new buffer where largest contiguous free 0 bytes)

Exception Backtrace:
--
1 | java.util.HashMap.createEntry | HashMap.java, line 826
2 | java.util.HashMap.putForCreate | HashMap.java, line 481
3 | java.util.HashMap.putAllForCreate | HashMap.java, line 486
4 | java.util.HashMap.<init> | HashMap.java, line 257
5 | android.app.SharedPreferencesImpl$EditorImpl.commitToMemory | SharedPreferencesImpl.java, line 402

6 | android.app.SharedPreferencesImpl$EditorImpl.apply | SharedPreferencesImpl.java, line 362
7 | com.pushwoosh.a.f.a | Unknown Source
8 | com.pushwoosh.notification.LocalNotificationReceiver.scheduleNotification | Unknown Source
9 | com.pushwoosh.notification.e.f | Unknown Source
10 | com.pushwoosh.g.a | Unknown Source
11 | com.pushwoosh.g.a | Unknown Source
12 | com.pushwoosh.j.onReceive | Unknown Source
13 | com.pushwoosh.internal.event.EventBus.b | Unknown Source
14 | com.pushwoosh.internal.event.EventBus.c | Unknown Source
15 | com.pushwoosh.internal.event.EventBus.a | Unknown Source
16 | com.pushwoosh.internal.event.c.run | Unknown Source
17 | android.os.Handler.handleCallback | Handler.java, line 751

18 | android.os.Handler.dispatchMessage | Handler.java, line 95
19 | android.os.Looper.loop | Looper.java, line 154
20 | android.app.ActivityThread.main | ActivityThread.java, line 6682
21 | java.lang.reflect.Method.invoke | Native Method
22 | com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run | ZygoteInit.java, line 1534
23 | com.android.internal.os.ZygoteInit.main | ZygoteInit.java, line 1424



```

**Comment by n-belokopytov** [comment]
Found one more.

`

Exception Backtrace:
--
1 | java.util.HashMap.inflateTable | HashMap.java, line 287
2 | java.util.HashMap.put | HashMap.java, line 419
3 | java.util.HashSet.add | HashSet.java, line 217
4 | java.util.AbstractCollection.addAll | AbstractCollection.java, line 342
5 | java.util.HashSet.<init> | HashSet.java, line 117
6 | com.pushwoosh.a.f.b | Unknown Source
7 | com.pushwoosh.a.f.a | Unknown Source
8 | com.pushwoosh.notification.LocalNotificationReceiver.scheduleNotification | Unknown Source
9 | com.pushwoosh.notification.e.f | Unknown Source
10 | com.pushwoosh.g.a | Unknown Source
11 | com.pushwoosh.g.a | Unknown Source
12 | com.pushwoosh.j.onReceive | Unknown Source
13 | com.pushwoosh.internal.event.EventBus.b | Unknown Source
14 | com.pushwoosh.internal.event.EventBus.c | Unknown Source
15 | com.pushwoosh.internal.event.EventBus.a | Unknown Source
16 | com.pushwoosh.internal.event.c.run | Unknown Source
17 | android.os.Handler.handleCallback | Handler.java, line 751

`

**Comment by wfhm** [comment]
@n-belokopytov the fix to OOM errors will be included in the next release.

**Comment by wfhm** [comment]
@n-belokopytov fixed in 5.4.0.
